Mc 1.17 lighting & Load Order

CursedFlames opened the pull request:
Draft PR to track progress.

Status as of making this PR: server-side lighting seems to work correctly, both during worldgen and when there are block updates. Obviously no handling of unloading/loading yet, since we've yet to do save file stuff. Client-side lighting is still broken. (copied from #529)

CursedFlames committed:
Begin to implement a skylight heightmap. Doesn't handle any client-side stuff yet.

CursedFlames committed:
misc cleanup, and make light heightmap update mark neighboring section as dirty when necessary

CursedFlames committed:
create and update light heightmap during worldgen. Currently awful for performance; needs optimization

CursedFlames committed:
disable vanilla skylight source handling

CursedFlames committed:
begin implementing server-side sky light

CursedFlames committed:
more WIP lighting stuff; still mostly non-functional

CursedFlames committed:
horrendous merge commit and some debug code

CursedFlames committed:
fix CubePrimers not being added to light heightmaps properly

CursedFlames committed:
slight code cleanup and isCubic checks

CursedFlames committed:
add light heightmap data to F3 menu

CursedFlames committed:
sync light heightmap updates to the client

CursedFlames committed:
add comments to distinguish local/global positions, and fix global positions being used in a few places that should use locals

CursedFlames committed:
Several lighting fixes. Lighting should be mostly functional now.

CursedFlames committed:
Merge branch 'MC_1.17' into MC_1.17_lighting

CursedFlames committed:
isCubic checks

CorgiTaco committed:
Merge branch 'MC_1.17' into MC_1.17_lighting

CursedFlames committed:
Fix skylight updates during cube generation. Still broken somehow though

CursedFlames committed:
add CubePrimers to CubeMap during lighting stage

CursedFlames committed:
pass checkstyle

CorgiTaco commented:
This PR almost definitely requires load order to be fixed in order to patch and fix the numerous light leaks that infrequently & inconsistently occur in world generation.

Niko-sk2x committed:
Fix full skylight in random places

CorgiTaco committed:
-Make ImposterProtoChunk use LevelChunk's CubeMap and LightHeightmap

CursedFlames committed:
misc cleanup

CursedFlames committed:
Merge branch 'MC_1.17' into MC_1.17_lighting

CursedFlames committed:
replace LayerLightEngine.getStateAndOpacity inject with a redirect

CursedFlames committed:
Merge branch 'MC_1.17' into MC_1.17_lighting

CursedFlames committed:
mixin to remove unnecessary skylight logic

CursedFlames committed:
fix no skylight in sections without skylight storage

CorgiTaco committed:
-Allow getComputedLevel to run on the client
-CHeck if a cube is either null or before light status

CursedFlames committed:
rename some lighting mixins to match mojang mappings

CursedFlames committed:
store light heightmap sections on CubePrimer and BigCube, and make light heightmap handle upgrade from CubePrimer to BigCube

CursedFlames committed:
implement loadLightHeightmapSection on BigCube too

CursedFlames committed:
Merge branch 'MC_1.17' into MC_1.17_lighting

CorgiTaco committed:
Merge branch 'MC_1.17' into MC_1.17_lighting

CorgiTaco committed:
-Fix client and server siding

CorgiTaco committed:
-Make SkyDataLayerStorageMap non protected

NotStirred committed:
Revert "[1.17] Columns before cubes load order (#671)"

NotStirred committed:
base load order + unload order

NotStirred committed:
Fix future chain order

NotStirred committed:
Fix debug status renderer displaying incorrect statuses for cubes & columns
 - was previously displaying target status, which were always invalid for chunks for load order

NotStirred committed:
attempted fix for https://bugs.mojang.com/browse/MC-224894 in CC code

CursedFlames committed:
Merge pull request #703 from OpenCubicChunks/MC_1.17_load_order_4

Columns before cubes load order & unload order

Cyclonit committed:
- Replaced warnings/error handling on accessing missing chunks with assertions. Load order should guarantee chunks being present.
- Renamed CubePrimer.loadLightHeightmapSection to CubePrimer.setLightHeightmapSection as that is what it does.

Cyclonit committed:
- Introduce proper lazy loading for heightmaps in CubePrimer.getHeightmapSections.

Cyclonit committed:
- Fixed light heightmaps being initialized in LIGHTING instead of FEATURES, resulting in missing heightmaps on the server.
- Removed assertions for missing chunks until client load order is fixed.
- Added forcing CCColumn tickets for every cube ticket to fix missing chunks.
- Removed unnecessary debugging code for lazily initialing light heightmaps.

CursedFlames committed:
fix skylight propagation only using horizontal directions

CursedFlames committed:
prevent vanilla behaviour in SkyLightSectionStorage.getLightValue on client

CursedFlames committed:
Merge branch 'MC_1.17' into MC_1.17_lighting

CursedFlames commented:
Lighting still isn't fully functional, but we probably want to merge this now instead of waiting until it's perfect. Server-side lighting seems to work pretty well, at least.

CursedFlames committed:
pass checkstyle

NotStirred commented:
I want to fix the CCColumn tickets before merge, as currently it has two solutions

NotStirred committed:
Properly add CCColumn tickets for all columns within a cube

CorgiTaco commented:
This behavior: https://github.com/OpenCubicChunks/CubicChunks/blob/c9575d54a1147bc25d8b2b82592e7fd7c0af53a2/src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/lighting/MixinSkyLightSectionStorage.java#L58


Is intended and should be handled accordingly in way that works without a chunk

CorgiTaco committed:
-Assert if the cube at the position is not empty and the chunk is null on client

CursedFlames commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/client/debug/MixinDebugScreenOverlay.java:
```diff
@@ -0,0 +1,57 @@
+            locals = LocalCapture.CAPTURE_FAILHARD)
+    private void onGetGameInformation(CallbackInfoReturnable<List<String>> cir, String string2, BlockPos pos, Entity entity, Direction direction,
+                                      String string7, Level level, LongSet longSet, List<String> list, LevelChunk clientChunk, int i) {
+        LevelChunk serverChunk = this.getServerChunk();
```
We still need to add proper integrated server checks here, since getServerChunk() can also return the client chunk for some reason (thanks Mojang)

CorgiTaco replied:
Added if checks for the integrated server prior to loading the server side data in the debug overlay

CorgiTaco committed:
-Check if the integrated server exists before adding server light data to the debug overlay

Cyclonit commented on build.gradle:
```diff
@@ -132,6 +132,7 @@ minecraft {
             "-Dcubicchunks.debug.window=false",
             "-Dcubicchunks.debug.statusrenderer=false",
             "-Dcubicchunks.debug.biomes=false",
+            "-ea"
```
Should this stay?

CursedFlames replied:
I figure we probably want to keep assertions on, to tell if there are any cases where load order breaks and such.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/BigCube.java:
```diff
@@ -218,10 +222,26 @@ public BigCube(Level worldIn, CubePrimer cubePrimer, @Nullable Consumer<BigCube>
         this.setAllReferences(cubePrimer.getAllReferences());
 //        var4 = protoChunk.getHeightmaps().iterator();
```
Can this go?

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/BigCube.java:
```diff
@@ -218,10 +222,26 @@ public BigCube(Level worldIn, CubePrimer cubePrimer, @Nullable Consumer<BigCube>
 //        var4 = protoChunk.getHeightmaps().iterator();
 
+        LightSurfaceTrackerSection[] primerLightHeightmaps = cubePrimer.getLightHeightmaps();
+        for (int i = 0; i < IBigCube.DIAMETER_IN_SECTIONS * IBigCube.DIAMETER_IN_SECTIONS; i++) {
```
IBigCube.CHUNK_COUNT

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/BigCube.java:
```diff
@@ -983,14 +1003,20 @@ public void postLoad() {
-                // TODO we really, *really* shouldn't be force-loading columns here.
-                //      probably the easiest approach until we get a "columns before cubes" invariant though.
+
+                // TODO force-loading columns is questionable, until we get load order
```
Is this force loading still necessary?

CursedFlames replied:
Load order should guarantee that the column is loaded already anyway, AFAIK.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/cube/CubePrimer.java:
```diff
@@ -166,78 +176,180 @@ public void setHeightToCubeBounds(boolean cubeBounds) {
+//
+//        if (toInitialize != null) {
+//            primeHeightMaps(toInitialize);
+//        }
```
This can go.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/heightmap/ClientSurfaceTracker.java:
```diff
@@ -10,13 +10,18 @@
+     * @param x column-local x
+     * @param y global y
+     * @param z column-local z
+     */
```
Is there any reason not to rename the parameters themselves?
public boolean update(int columnLocalX, int globalY, int columnLocalZ, BlockState blockState)

CursedFlames replied:
looks a bit messier IMO, but sure

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/heightmap/SurfaceTrackerSection.java:
```diff
@@ -252,7 +255,8 @@ private SurfaceTrackerSection loadNode(int newScaledY, int sectionScale, IBigCub
     }
 
-    private int index(int x, int z) {
+    /** Get position x/z index within a column, from global/local pos */
```
Isn't there a helper method for this elsewhere?  There should be.

CursedFlames replied:
Not seeing one in any obvious place (`Coords`, etc)

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/heightmap/SurfaceTrackerWrapper.java:
```diff
@@ -22,14 +24,30 @@ public SurfaceTrackerWrapper(ChunkAccess chunkAccess, Types types) {
+     * @param blockState unused.
+     * @return currently unused; always false
+     */
     @Override
```
Any reason not to rename the parameters instead of adding the java doc?

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/heightmap/SurfaceTrackerWrapper.java:
```diff
@@ -22,14 +24,30 @@ public SurfaceTrackerWrapper(ChunkAccess chunkAccess, Types types) {
         return false;
     }
 
+    /** x/z are column-local. */
```
rename x and z to localX, localZ

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/MixinMinecraftServer.java:
```diff
@@ -68,10 +68,10 @@ private void prepareLevels(ChunkProgressListener statusListener, CallbackInfo ci
         int d = radius * 2 + 1;
         ((IServerChunkProvider) serverchunkprovider).addCubeRegionTicket(TicketType.START, spawnPosCube, radius + 1, Unit.INSTANCE);
-        serverchunkprovider.addRegionTicket(TicketType.START, spawnPosCube.asChunkPos(), Coords.cubeToSection(radius + 1, 0), Unit.INSTANCE);
+//        serverchunkprovider.addRegionTicket(TicketType.START, spawnPosCube.asChunkPos(), Coords.cubeToSection(radius + 1, 0), Unit.INSTANCE);
```
Can this go?

NotStirred replied:
yes

Niko-sk2x replied:
Did we get to the point of this not being necessary?

NotStirred replied:
This has been replaced by cubes adding CCColumn tickets in `MixinTicketManager#addCubeTicket`

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/MixinMinecraftServer.java:
```diff
@@ -68,10 +68,10 @@ private void prepareLevels(ChunkProgressListener statusListener, CallbackInfo ci
-        while (this.isRunning() && (serverchunkprovider.getTickingGenerated() < chunkDiameter * chunkDiameter
-            || ((IServerChunkProvider) serverchunkprovider).getTickingGeneratedCubes() < d * d * d)) {
+        while (this.isRunning() && (/*serverchunkprovider.getTickingGenerated() < chunkDiameter * chunkDiameter
+            ||*/ ((IServerChunkProvider) serverchunkprovider).getTickingGeneratedCubes() < d * d * d)) {
```
Can this go?

NotStirred replied:
yes

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java:
```diff
@@ -84,6 +110,52 @@ private void onInit(Level levelIn, ChunkPos pos, ChunkBiomeContainer chunkBiomeC
+        if (!this.isCubic()) {
+            return;
+        }
+        // TODO client side light heightmap stuff
```
What kind of stuff?

CursedFlames replied:
Still not sure what needs to be done with the client light heightmap, in terms of updating when there are client block updates vs. receiving updates from the server. Doesn't really matter for this PR, since we just want server-side light working for now. I'll make the comment clearer about what might need to be done though.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinChunk.java:
```diff
@@ -84,6 +110,52 @@ private void onInit(Level levelIn, ChunkPos pos, ChunkBiomeContainer chunkBiomeC
+        } else {
+            lightHeightmap = new LightSurfaceTrackerWrapper(this);
+        }
+        // TODO might want 4 columns that share the same BigCubes to have a reference to the same CubeMap?
```
Afaik @Barteks2x wanted something along these lines. Could we check if any other relevant chunk exists and use their CubeMaps or would this break something?

CursedFlames replied:
This might be possible, but it would be kinda messy. Probably best to leave it as-is for now, and consider changing it later.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/chunk/MixinProtoChunk.java:
```diff
@@ -104,4 +130,18 @@ private void setMinHeight(CallbackInfoReturnable<Integer> cir) {
+        if (!this.isCubic()) {
+            return;
+        }
+        if (lightHeightmap == null && this.getStatus().isOrAfter(ChunkStatus.FEATURES)) {
```
Do we want to initialize these earlier on? Barteks mentioned that it might improve performance.

Cyclonit replied:
As this is not an issue for now, we should leave it as is and address it in the future.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/entity/MixinServerPlayerEntity.java:
```diff
@@ -21,4 +29,13 @@ public void onSendChunkLoad(ServerGamePacketListenerImpl serverPlayNetHandler, P
+//        if (((CubicLevelHeightAccessor) this.getLevel()).isCubic()) {
+//            PacketDispatcher.sendTo(new PacketUpdateLight(CubePos.from(new BlockPos(this.position())), this.level.getLightEngine(), true), (ServerPlayer) (Object) this);
+//        }
+//    }
```
Remove this pre-merge.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/lighting/MixinLayerLightSectionStorage.java:
```diff
@@ -78,13 +77,15 @@ protected void markNewInconsistenciesForCube(LayerLightEngine<M, ?> engine, bool
-                if (this.cubesToRetain.contains(CubePos.sectionToCubeSectionLong(noLightPos))) {
+                // FIXME this commented out check is probably important, but also breaks client-side lighting
+                //       should investigate why it breaks things instead of just disabling it.
+//                if (this.cubesToRetain.contains(CubePos.sectionToCubeSectionLong(noLightPos))) {
```
Is this comment up to date? Can we uncomment the check?

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/lighting/MixinLayerLightSectionStorage.java:
```diff
@@ -99,14 +100,16 @@ protected void markNewInconsistenciesForCube(LayerLightEngine<M, ?> engine, bool
-                if (this.storingLightForSection(entryPos)) {
+                // FIXME this commented out check is also probably important, but this one breaks *server-side* lighting
+                //       should investigate why it breaks things instead of just disabling it.
+//                if (this.storingLightForSection(entryPos)) {
```
Can this be reenabled?

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/lighting/MixinSkyLightEngine.java:
```diff
@@ -0,0 +1,223 @@
+            ((LayerLightSectionStorageAccess) this.storage).invokeRunAllUpdates();
+
+            // TODO cube iteration order might still be important here
+            //      (int y = oldHeight-1; y >= newHeight; y--)
```
Is it?

CursedFlames replied:
Since light updates are queued before being processed, it shouldn't affect it, but I'm not entirely confident about that.

Cyclonit replied:
We currently don't see any lighting errors. I'd vote for removing that ToDo.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/lighting/MixinSkyLightSectionStorage.java:
```diff
@@ -0,0 +1,144 @@
+
+            if (chunk == null) {
+                // ToDo Known bug: Cubes may be sent to the client ahead of their chunks. This is not fatal.
+                // see MixinSkyLightEngine.onGetComputedLevel(...)
```
This should no longer be the case.

Cyclonit commented on src/main/java/io/github/opencubicchunks/cubicchunks/mixin/debug/common/MixinMinecraftServer.java:
```diff
@@ -90,11 +90,11 @@ private void prepareLevels(ChunkProgressListener worldGenerationProgressListener
+//            Thread.sleep(10000);
+//        } catch (InterruptedException e) {
+//            int ignored = 0;
+//        }
```
This should not be commented out. It should be disabled through the mixin.

NotStirred replied:
if the `cubicchunks.debug` jvm arg is not set this mixin will never apply anyway.
This was here to wait on loading into the world to make sure all of the cubes were properly generated

CorgiTaco committed:
Merge branch 'MC_1.17' into MC_1.17_lighting

CursedFlames committed:
address simple PR reviews

NotStirred committed:
properly convert region ticket level to chunk space when adding a cube region ticket

NotStirred committed:
use variable cube size for heightmap array size

NotStirred committed:
cube region tickets do not need to have equivalent chunk region tickets added due to tickets added in cube generation stages
This reverts commit 33973a6a

NotStirred committed:
Fix inverted load order assert check

NotStirred committed:
remove outdated todos, added extra asserts if a chunk is null

NotStirred committed:
Minor refactor of ticket manager + checkstyle

Cyclonit committed:
Reenable retaining of light data for cubes.
Always ignore missing lighting information for rendering of entities.

Cyclonit commented:
I consider lighting to be working fine. I did further testing today and could not find a single lighting error.

Cyclonit commented:
> 
> 
> This behavior:
> 
> https://github.com/OpenCubicChunks/CubicChunks/blob/c9575d54a1147bc25d8b2b82592e7fd7c0af53a2/src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/lighting/MixinSkyLightSectionStorage.java#L58
> 
> Is intended and should be handled accordingly in way that works without a chunk



> 
> 
> This behavior:
> 
> https://github.com/OpenCubicChunks/CubicChunks/blob/c9575d54a1147bc25d8b2b82592e7fd7c0af53a2/src/main/java/io/github/opencubicchunks/cubicchunks/mixin/core/common/world/lighting/MixinSkyLightSectionStorage.java#L58
> 
> Is intended and should be handled accordingly in way that works without a chunk

What makes you say that this behaviour is intended? It definitely isn't as far as CC is concerned. We know the reason, see https://github.com/OpenCubicChunks/CubicChunks/issues/728

Niko-sk2x commented on src/main/java/io/github/opencubicchunks/cubicchunks/chunk/IChunkManager.java:
```diff
@@ -121,6 +122,4 @@ static int getCubeDistanceXZ(CubePos cubePosIn, int x, int z) {
 
     boolean noPlayersCloseForSpawning(CubePos cubePos);
-
-    void setServerChunkCache(ServerChunkCache cache);
```
Why is this method moved?

NotStirred commented:
I cant respond to https://github.com/OpenCubicChunks/CubicChunks/pull/636#discussion_r685594006 as it seems to be a part of an already resolved conversation. Is this a github bug?

This has been replaced by cubes adding CCColumn tickets in `MixinTicketManager#addCubeTicket`
